### DISCUSSION TITLE
OptimizationReporter

### DISCUSSION THREAD

**Discussion by am-tc01** [idaholab/moose#23772] [Q&A Modules: General] [answered]
In one of the [optimization tests](https://github.com/idaholab/moose/blob/next/modules/optimization/test/tests/optimizationreporter/bimaterial/main.i), why is there one value in `lower_bounds` and `upper_bounds`, while `num_values = 2`. What if I want different lower and upper bounds for each value in `diffusivty_values`?

```
[OptimizationReporter]
  type = OptimizationReporter
  parameter_names = diffusivity_values
  num_values = 2 # diffusivity in the bottom material and in the top material of model.i
  initial_condition = '3 4' # the expected result is about '1 10' so this initial condition is not too bad
  lower_bounds = '1'
  upper_bounds = '20'
  measurement_file = 'synthetic_data.csv'
  file_value = 'temperature'
[]
```

Regards.

**Comment by GiudGiud** [comment]
Hello

This would be an optimization that has not been added to the code.
Feel free to add it, we would just need to have a "vector of vector" syntax for the parameter, then get the indexing right

Guillaume

**Reply by GiudGiud** [reply]
@zachmprince for awareness

**Reply by am-tc01** [reply]
Thank you for your answer.
In an earlier version, one could just specify the name of the vector in `parameter_names` and the size of the vector in `num_values`, and then specify `lower_bounds `and `upper_bounds `for the same vector length as `num_values`. But now, I think, this is changed, and it gets somehow a little confusing. The following worked in an earlier version, but now it gives an error: `There must be a lower bound associated with each parameter.`
```
[OptimizationReporter]
  type = OptimizationReporter
  parameter_names = diffusivity_values
  num_values = 2 # diffusivity in the bottom material and in the top material of model.i
  initial_condition = '3 4' # the expected result is about '1 10' so this initial condition is not too bad
  lower_bounds = '1 2'
  upper_bounds = '10 20'
  measurement_file = 'synthetic_data.csv'
  file_value = 'temperature'
[]
```

**Reply by lynnmunday** [reply]
@am-tc01   I removed the ability to have different lower and upper bounds for each parameter.  I wanted to make the optimizationReporter consistent with the mesh based method for defining parameters, ParameterMeshOptimization.  Do you have a good reason for needing different lower and upper bounds for the same material property?  Maybe it is something I can try to add back in.

**Reply by am-tc01** [reply]
@lynnmunday thanks for your answer.
One could have different properties or different parameters to optimize for. For instance, in my case I have 5 different parameters to optimize for and then I need to have them in the same `OptimizationReporter`. It's not logical to have same upper and lower bounds on all those. Not sure how one can have a vector of vectors, have different upper and lower bounds on each and then transfer them to a sub-app?

**Reply by zachmprince** [reply]
@am-tc01 Just so you know, you can have multiple parameter groups. So you can change your input to have:
```
[OptimizationReporter]
  type = OptimizationReporter
  parameter_names = 'diffusivity_bottom diffusivity_top'
  num_values = '1 1' # diffusivity in the bottom material and in the top material of model.i
  initial_condition = '3 4' # the expected result is about '1 10' so this initial condition is not too bad
  lower_bounds = '1 2'
  upper_bounds = '10 20'
  measurement_file = 'synthetic_data.csv'
  file_value = 'temperature'
[]
```

There is an example of doing this [here](https://github.com/idaholab/moose/blob/next/modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/main.i).
It seems like this might be cumbersome for your problem though. We can consider adding the capability to have both options: one for single bound for all values, and one for a bound for each value.

**Reply by am-tc01** [reply]
@zachmprince Yes that is correct. In principle this can be done but then if one has many parameters, it could get very cumbersome to deal with all the parameters separately in the sub-app. I would really appreciate if you could add this capability again, to have a bound for each value.

Regards.

**Reply by lynnmunday** [reply, marked as answer]
@am-tc01 I added this pull request https://github.com/idaholab/moose/pull/23868 to change this and make all of these parameters a vector of vectors.  The input syntax changed a bit to do this, you now need to use a ";"  to separate each vector of parameters.  You can see this in some changes to input files in that pr.

**Reply by am-tc01** [reply]
@lynnmunday Super! many thanks!

**Reply by lynnmunday** [reply]
@am-tc01 could you tell me anything about the problems your are trying to solve with the optimization module?  I'm just curious how its being used.

**Reply by am-tc01** [reply]
@lynnmunday I am trying to optimize processing parameters for certain manufacturing processes.
I wonder if you or others, if you know, have used optimization module for some rigorous optimization problems? If yes, then how, in general is your experience with Petsc TAO solvers?

**Reply by lynnmunday** [reply]
@am-tc01   The PR was merged so you should be able to do what you want with it now.  For linear force inversion problems, TAO works great.  There is an example for a time domain force inversion problem where the forces are being optimized at ever time step resulting in 1000's of tunable parameters and tao bfgs works fine.  TAO works as well as it can for material inversion problems.   There are convergence issues with material inversion problems but I think that is because we are trying to optimize too many material parameters in the domain and the measurement points and forcing points do not produce any sensitivity in some of the domains where we are trying to invert for the material property.  Regularization would help us get convergence in these cases but it wouldn't be correct.  For material inversion, you really need a good initial guess.

**Reply by am-tc01** [reply]
@lynnmunday Thanks for your insight! I agree with you. Tao bfgs works well for me if I go with a warm start and most often doesn't converge if I use some random initial values. 
I'll update and try in a day or two. Thanks! This thread can be closed.
